### DISCUSSION TITLE
building DIY zigbee sensors

### DISCUSSION THREAD

**Discussion by martin3000** [zigpy/zigpy#867] [General]
I have built many sensors with Esp8266 and Esp32 modules. They are easy to program with the arduino IDE und easy to flash with USB and a serial boot loader. 
Are there similar modules for zigbee which can also be flashed via USB? PTVO does not support serial flashing.
Is it possible to use libraries for zigbee, so that I can use something like "zigbeeSendData(temperature)" ?

**Comment by MattWestb** [comment]
My opinion is that CC-253X is very underpowered and is using old Zigbee HA 1.x stack (not Zigbee PRO) and having many bad sides so no alternative for my.

I have doing one [ESP-NTP-DCF77-Encoder](https://github.com/MattWestb/ESP-NTP-DCF77-Encoder) that is using WIFi and coding DCF77 for injecting the signal in radio watches and also one 77 Mhz modulator for "sending" real signals if the decoder is integrated in the MCU and also one not publick weather station with arduino that getting all data from MQTT from HA and it working good.

But for small nice things i using ESPHome that is easier to using then not good in programming but i preferring Zigbee then its possible.

The bet if you is good in programming is using one Silabs WSTK kit and getting license on ther Zigbee SDK but its costing 100 € but then you is getting one very good onboard J-Link programmer plus one or 2 radio test boards and al tools you can using for building real Zigbee devices only need doing the coding right.

**Comment by Adminiuga** [comment]
There are modules that are very easy to flash. It's just an USB mass storage device to which you copy the firmware. One is the Thenderboard Sense 2 module and it already has temperature, humidity and illuminance and bunch of other sensors. 
https://www.digikey.com/en/product-highlight/s/silicon-laboratories/thunderboard-sense-2-next-generation-iot-development-kit

**Reply by Adminiuga** [reply]
It is not possible to use zigpy on this module.

**Reply by martin3000** [reply]
wow, that sounds great! I have ordered one!
what is the best starting point for a begginer? how to send "hello world" over zigbee?

**Reply by Adminiuga** [reply]
Over ZigBee it is not so easy. As you do need to have access to EmberZNet SDK which is not free (the cheapest bone is $100) but BLE SDK is free and you can send data over Bluetooth or custom protocols. 

EmberZNet SDK has some application example which are good starting point. There are also some boot camps on GitHub for ZigBee EmberZNet app framework

**Reply by martin3000** [reply]
if I use wifi on an esp32, I do wifi.begin(), wifi.connect() and so on and that's it. I don't need a SDK. How does ptvo work?

**Reply by Hedda** [reply]
Note that PTVO's Zigbee Configurable Firmware tool only builds firmware for Texas Instruments CC253x, CC2x52, and CC1x52 series.

> How does ptvo work?

With PTVO's Zigbee Configurable Firmware tool you basically configure the settings you want in it and then it builds a firmware image. 

As such, using PTVO's Zigbee Configurable Firmware tool does not require any Zigbee experience or even great coding expertice. 

> PTVO does not support serial flashing.

You can not flash with the PTVO tool itself as instead you can flash that firmware image in whatever way you want / can.

That is, you can flash PTVO generated firmware in the same way you would flash https://github.com/Koenkk/Z-Stack-firmware/

**Reply by martin3000** [reply]
I flashed a cc2652 stick with PTVO via USB and afterwards I was told that ptvo does not support a serial bootloader. So the stick is bricked now.

**Reply by glcos** [reply]
>  is bricked now.

@martin3000 there are chances that your stick is not bricked, you just need to reflash it with a JTAG programmer such as Texas Instruments XDS110.
I've been playing with PTVO firmware for quite a while and always used a TI programmer to flash devices.

**Reply by martin3000** [reply]
yes ok, but then I have to order j-tag programmer and experiment the whole afternoon.

**Reply by glcos** [reply]
> yes ok, but then I have to order j-tag programmer and experiment the whole afternoon.

In case it helps, I have the Chinese clone purchased on AliExpress and it works well, it also updates its own firmware when attached to TI Flash programmer application.

**Reply by Hedda** [reply]
Note! Official CC2652 devlopment kit and CC1352 development kit (a.k.a. TI's SimpleLink LaunchPad dev kits) are not super expensive:

https://www.ti.com/design-resources/embedded-development/hardware-kits-boards.html

https://www.ti.com/tool/LPSTK-CC1352R

https://www.ti.com/tool/LP-CC2652RB

https://www.ti.com/tool/LAUNCHXL-CC1352P

> In case it helps, I have the Chinese clone purchased on AliExpress and it works well, it also updates its own firmware when attached to TI Flash programmer application.

Do you have a link?

**Reply by martin3000** [reply]
> As you do need to have access to EmberZNet SDK which is not free

It is free if you have the device (23€). It is installed automatically in simplicity studio.

**Reply by MattWestb** [reply]
Did you getting access granted the registering the Thenderboard Sense 2 board to EZSP (Zigbee SDK) ?
I like the board but have not buying it then its have very much sensors onboard and can being customized if having Zigbee stack access and also have on onboard J-Link ODB that is great and also PTS for sniffing Zigbee traffic "in real".

**Reply by martin3000** [reply]
I installed Simplicity Studio, plugged the thunderboard in the USB and all SDKs were installed. The Simplicity Studio tests the device. Without the board, you cannot install anything.

**Reply by Hedda** [reply]
> I installed Simplicity Studio, plugged the thunderboard in the USB and all SDKs were installed. The Simplicity Studio tests the device. Without the board, you cannot install anything.

Nice! 

So buying [Silicon Labs Thunderboard Sense 2 IoT Dev Kit (SLTB004A)](https://www.silabs.com/development-tools/thunderboard/thunderboard-sense-two-kit) lowers minimum entry to EmberZNet Zigbee SDK to only ~$20?

I thought the minimum was to buy [Silabs EFR32xG22 Wireless Gecko Starter Kit (SLWSTK6021A)](https://www.silabs.com/development-tools/wireless/efr32xg22-wireless-starter-kit) for ~$99 to get Zigbee SDK access?

**Reply by martin3000** [reply]
> So buying [Silicon Labs Thunderboard Sense 2 IoT Dev Kit (SLTB004A)](https://www.silabs.com/development-tools/thunderboard/thunderboard-sense-two-kit) lowers minimum entry to EmberZNet Zigbee SDK to only ~$20?

Maybe you can use the SDK only with simplicity studio for that device!?

**Reply by Hedda** [reply]
> Maybe you can use the SDK only with simplicity studio for that device!?

Pretty sure access to Silabs EmberZNet Zigbee SDK is all-of-nothing (as in either you have access to everything Zigbee or you do not?)

**Reply by martin3000** [reply]
here it is for free: https://github.com/SiliconLabs/gecko_sdk/releases

**Reply by Adminiuga** [reply]
Did you get EmberZNet SDK installed?

**Reply by MattWestb** [reply]
I was updating SS5 and have getting the starting guide for Gecko SDK (GSDK) Version 4.0.0.0 but the Zigbee EmberZNet SDK 7.0.0.0 GA is not showing up and it shall being released so somthing spooky is going on with SS5.
Must reading more for finding some new bugs and fixes that can being good to knowing.

**Reply by MattWestb** [reply]
@Adminiuga Something for you in the future **NCP Memory Allocation**  😄

**Reply by MattWestb** [reply]
I was downloading the Gecko SDK (GSDK) Version 4.0.0.0 from the new GIT and putting it in one new folder and importing it in SS5 and its loading it but complaining its not trusted but its installing OK.
NCP SW is **not** supported in GA its coming later so no worry for EZSP 7.0.0.0 on "normal" hardware !!!
Also most is only SDK updates and very little fixes for devices if reading the change log.
Have not testing compiling so i dont knowing if all is working OK.

**Reply by martin3000** [reply]
it is not necessary to install EmberZnet SDK 7, it is installed together with Gecko 4. But maybe you need an usb attached silabs device.

**Reply by MattWestb** [reply]
I have one WSTK with full licence so it shall updating all parts but its looks silabs have missing some part putting all things in old installed  SS5.
The WSTK is also attached but the licence is with login to Silabs for running the Zigbee stack part.

**Reply by Adminiuga** [reply]
well, gecko SDK 4.0 is on Github https://github.com/SiliconLabs/gecko_sdk/ Does this mean you don't need to purchase the WSTK at all?
Is the sky falling or did Silabs finally allows free access to EmberZNet?

**Reply by Adminiuga** [reply]
Can't get an example app compiled. For some reason the object like `/platform/radio/rail_lib/autogen/librail_release/librail_efr32xg12_gcc_release.a` are ascii files like
```
version https://git-lfs.github.com/spec/v1
oid sha256:1ffed2d64cecaae8e4be2aa7319cd41a518433ff44b9b2e1c98e0f8af11c6175
size 462802
```

Instead of the actual archives. Why? the actual git repo has correct files.

**Reply by Adminiuga** [reply]
nvm, need `git-lfs`, after git-lfs fetch and git-lfs pull the example project compiled. Exciting. 

I actually think this actually is going to improve the app framework greatly. Cause some parts out of the box in the previous version were hit or miss. For example the "reporting" plugin in EmberZNet  6.8 does not support "small" float values.

**Reply by MattWestb** [reply]
Martin was writing hi must have the thunderboard connected so it can being they is only checking its one (Silabs) J-Link OBD connected and can using SDK V4. But hi cant getting access to the old SDKs then needs login and having access for it.
I was doing it "user way" by downloading the Zip from the git and unzipping it in one parallel folder with 3.x and it looks working and i is having your missed file in it.
Shall looking if i can using some "extra" parts that is not Zigbee like BT mesh that need extra access.

**Reply by martin3000** [reply]
> the example project compiled. Exciting.

which example did you test? Z3Switch?
It was not possible to migrate SS4 examples with EmberZnet 6.

**Comment by Hedda** [comment]
Referred martin3000 to post here so copying his original question about DIY Zigbee devices and reply from https://github.com/zigpy/bellows/issues/348 for reference:

> something offtopic: are there modules which allow the creation of DIY zigbee sensors?

Yes there are. Texas Instruments CC2530 based radio modules from Ebyte and RF-star was previously most popular for DIY project until recently when more hobbyist started moving over to their newer modules based on CC2652 series (CC2650R/CC2650RB or CC2652P). 

Examples modules are RF-BM-2652P1 / RF-BM-2652P1 and RF-BM-2652P2 modules from RF-Star, and E72-2G4M20S1E module from Ebyte.

https://www.szrfstar.com/product/list-196-en.html

https://www.szrfstar.com/product/189-en.html

https://www.szrfstar.com/product/236-en.html

https://www.szrfstar.com/product/237-en.html

https://www.ebyte.com/en/product-view-news.html?id=1002

What drives most those projects Zigbee Configurable Firmware tool by PTVO -> https://ptvo.info as it allows just about anyone to build firmware for Texas Instruments based radio modules -> https://ptvo.info/zigbee-configurable-firmware-features/ and that recently also added support for CC2652 and CC1352:

https://ptvo.info/zigbee-switch-configurable-firmware-v2-210/

https://ptvo.info/cc2652r1-cc2652rb-cc2652p-cc1352p2-configurable-router-firmware-748/

I think the most active community around such DIY Zigbee products is "DIYRuZ" at https://github.com/diyruz/ which I believe post most on https://modkam.ru/ and they have several open source projects posted for different Zigbee sensors. Checkout examples:

https://www.zigbee2mqtt.io/supported-devices/#v=DIYRuZ

Also checkout some cool Zigbee sensors from smartboxchannel -> https://github.com/smartboxchannel?tab=repositories

PS: smartboxchannel also sells those Zigbee devices prebuilt on Tindie under his Efektalab brand -> https://www.tindie.com/stores/diyberk/

**Reply by Hedda** [reply]
See that PTVO also started selling an optional premium version -> https://ptvo.info/zigbee-configurable-firmware-features/premium/

**Reply by martin3000** [reply]
what I need is a 12V monitor for my car.

**Reply by Hedda** [reply]
I guess if you can use PTVO's tool or not it depends more on what type of specific sensors you need as other than only supporting Texas Instruments Zigbee chips it is not open source, so if the sensors you want is not already supported by  PTVO then you are out of luck:

https://ptvo.info/zigbee-configurable-firmware-features/

**Reply by martin3000** [reply]
working with simplicity studio 5 and gecko sdk is very complex und difficult. I would not recommend it. this is something for people doing that as a fulltime job the whole day. you need months to get used to it. and all examples from the internet do not work because they are for previous versions.

**Reply by MattWestb** [reply]
You can using SS4 its have more examples and downloading different SDKs but then you is getting older Zigbee stacks but can being good if you like having some examples like RGB lights and or motion sensor.
I think early / middle  EZSP 6.X is having most working examples.

**Reply by martin3000** [reply]
with ESP32 and Arduino-IDE you have one file, you start with wifi.begin(), print("xyz") and some more lines and the thing is working. With simplicity studio you need 500 files just to write "hello" to the log.

**Reply by Adminiuga** [reply]
> with ESP32 and Arduino-IDE you have one file, you start with wifi.begin(), print("xyz") 

Only because someone hide all those complexities for you and gave you a nice wrapper for wifi.begin() to use. Do the same the MG21 and the next person will say thank you

**Reply by martin3000** [reply]
you are right of course. but I have the feeling that no new customers choose emberznet and simplicity studio 5.

**Reply by MattWestb** [reply]
One large "old" is IKEA and new is Philips HUE and also tuya is using Silabs chips and but IKEA looks using 6.5.x for old device and now is not known (MG21 devices).
tuya is using ZVL8 so they must using SS5 then its not implanted on older SDKs.

SS shall being one of the easiest SDK getting up and running comparing with most other Zigbee SGKs.
I have only making some NCPs and also some Zigbee Control Bridges for sending  commands to devices in the network but i is no procramer more copy and paste person.

**Reply by abmantis** [reply]
@MattWestb are you using the IKEA module with custom firmware?

**Reply by MattWestb** [reply]
@abmantis I was starting using them a coordinator with Tasmota and later ZHA and its working great after flashing one NCP firmware on them. I also experimenting with using them with Zigbee Control Bridge (it can sending and receiving commands in the network with CLI = the same mode IKEA is using in there GW).
For coding Switches i have not so much time for doing it and need more time and code knowledge. I think **Adminiuga** have doing one modded tuya switch that is working and is one tuya module.
I like making one double dimmer switch (E1743 * 2) of one module with on/off and dimmer up/down * 2.
Also one 3 or 4 gang dimmer with one button / group for toggle / dimmering but its needs much coding :-(

**Reply by abmantis** [reply]
My interest was using one as a custom 2 relay switch.
Which is the tuya one?

**Reply by martin3000** [reply]
Ewelink: https://de.aliexpress.com/item/1005002557411111.html

**Reply by MattWestb** [reply]
The "black" module is very likely one tuya module with MG1B chip (IKEA is using MG1P chips in 1 and second gen devices) and is working with one simple BMP / SWD probe for programming but the white is one MG21 chip (very good one that i using as coordinator from one LIDL LED strips controller) and is the same chip IKEA is using in the new 3 gen devices and need one reaal J-Link for programming $€$ (or working clones of J-Link).

**Reply by tube0013** [reply]
> Ewelink: https://de.aliexpress.com/item/1005002557411111.html

I bought one of these and have a 1 and 2 channel version already (they have been ewelink based on cc2531s I think. Haven't tried the 4 channel one yet.

**Reply by Hedda** [reply]
> > Ewelink: https://de.aliexpress.com/item/1005002557411111.html
> 
> I bought one of these and have a 1 and 2 channel version already (they have been ewelink based on cc2531s I think. Haven't tried the 4 channel one yet.

The 4-relays model says it supports "Zigbee 3.0" so strange if it is based on CC2530 or CC2631 as it don't officially support Zigbee 3.0?

https://aliexpress.com/item/1005002557411111.html

Picture of Zigbee module looks to read 4-relay model "ZS3L", if so that is a known Tuya development module based on EFR32MG21:

https://developer.tuya.com/en/docs/iot/zs3l?id=K97r37j19f496

https://fccid.io/2ANDL-ZS3L

![image](https://user-images.githubusercontent.com/6320001/148521664-59f8e3da-e06b-4941-b761-b493d87f1dec.png)

ZS3L Zigbee 3.0 module is also used in Tuya Sandwich Zigbee SoC Board (ZS3L)

https://developer.tuya.com/en/docs/iot/tuya-sandwich?id=K9lmlvkqwo6rn

![image](https://user-images.githubusercontent.com/6320001/148521414-33b54816-8c42-44ae-8657-83f92126ab73.png)

But picture of 1-relay and 2-relays models at least look to read "SM-015 V1.0" which I guess is CC2530 Zigbee module from Coolkit?

https://aliexpress.com/item/1005002364339462.html

https://www.coolkit.cn/product/sm-015/

![image](https://user-images.githubusercontent.com/6320001/148521158-81b66aad-e8ae-4ed5-a458-7897eda62ffe.png)

**Reply by MattWestb** [reply]
The ZS3L is one great module and its calibrated from the factory that the famous SM-011 is not and need setting corestion for the crystal in the firmware then is not calibrated in the factory (burned in the device config).
I think its in the same class as the original Silabs MG21 modules but is cheaper if can baying them on the open market. I'm using one LIDL / tuya LED Strip controller as coordinator on my test network with EZSP 6.10 :-))

The TI CC-2530/2531 Zigbee3 stack **is** certified but not updated then TI have stop supporting it https://zigbeealliance.org/zigbee_products/z-stack/ and in the end is one old and underpowered chip so better going with Silabs chips if like doing some more serious that have active support and not need being MG2X chips then MG1X is still working very good if using EZSP 6.7 stack but is still partly supported in EZSP 7.0 (not all is deleted in the last SDK then the large chips is not EOL).

The large problem is now getting the chips / modules then tuya and IKEA is having "little" demand for them if the can getting there logistic working then its broken all over the world and the small one like our Gary and other is not getting and parts to working with.

Easiest and cheapest getting good MG21 modules is baying the new IKEA lights for 8€ if not getting some other 3 gen product cheaper (our danich IKEA remote hackers have starting doing that and flashing them with Styrbar firmware).

**Comment by Hedda** [comment]
FYI, in related news, while there is no specific device news, Espressif announced ESP32-H2 SoC with Zigbee support, thought there are no releases of Zigbee stack / SDK / libraries  /code for it as of yet so nothing useful so far but probably worth keeping an eye on for the future:

https://www.cnx-software.com/2021/08/03/esp32-h2-risc-v-wisoc-announced-with-zigbee-3-thread-and-bluetooth-le-5-2/

https://www.espressif.com/en/news/ESP32_H2

**Reply by martin3000** [reply]
wow, that sounds great!

**Comment by glcos** [comment]
@Hedda @martin3000 
For whose who asked, I bought my XDS110 clone here
https://it.aliexpress.com/item/4000751867419.html

Looks like the device is being seen as a genuine Texas Instrument probe as you can see in the pic below

<img width="549" alt="xds110" src="https://user-images.githubusercontent.com/22558340/146808618-f00d5e03-23af-4f7f-86cc-95467a99b877.PNG">

Below how it looks like

![XDS110](https://user-images.githubusercontent.com/22558340/146808679-5aaf5ed1-8bd8-4a83-bb99-52e0efdc1a9c.jpg)

Hope this helps

**Reply by martin3000** [reply]
thank you

**Comment by martin3000** [comment]
With the silabs thunderboard sense 2 I have created a zigbee device with a few measurement sensors (clusters). When I pair it with ZHA, all clusters appear as sensor entities, except the dc voltage measurement on cluster 0x0b04. Is this not defined in ZHA?

**Reply by dmulcahey** [reply]
It’s not. Do you plan on sharing the project? I wouldn’t mind poking around it in my free time.

**Reply by Adminiuga** [reply]
DC voltage is not defined in zha. Are there any production device which report DC voltage? 

Can you modify that example to report voltage on power configuration cluster?

**Reply by martin3000** [reply]
> Can you modify that example to report voltage on power configuration cluster?

I could modify but I thought power config cluster is for the internal power supply of the device?
I want to build a power monitor for my car.

**Reply by martin3000** [reply]
> Do you plan on sharing the project? 

Its not yet a project. Its stabbing in the dark :-(

**Reply by dmulcahey** [reply]
That’s how projects start 😀

**Reply by Adminiuga** [reply]
> I want to build a power monitor for my car.

Yeah, in this case you are correct, it should be DC Voltage attribute. Get it working and adding it to ZHA shouldn't be a problem.

**Reply by martin3000** [reply]
> Can you modify that example to report voltage on power configuration cluster?

I modified it and it has got a power configuration cluster now, but in HA there is no battery sensor. Maybe because power source is set to mains?

**Comment by martin3000** [comment]
ZHA shows the following info for my new device:
```
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.EndDevice: 2>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, 
  frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.AllocateAddress|RxOnWhenIdle|MainsPowered: 140>, 
  manufacturer_code=4098, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, 
  descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, 
  *is_end_device=True, *is_full_function_device=False, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=False, *is_security_capable=False)",

```
Where does this info come from? MainsPowered: 140?

**Reply by Adminiuga** [reply]
It is in the node descriptor. Need to find out where it is configured in SDK, somewhere about ZigBee device configuration.

But better implement as DC voltage and add to zha a new sensor.

**Reply by Adminiuga** [reply]
So looks like you may need to go to SW Components -> Zigbee -> Utility -> Zigbee Device Config and configure device type as "End Device", this should change the Node Descriptor mains powered options. This requires you to uninstall the "Pro Zigbee Stack" and install the Leaf Zigbee stack.

Still need to figure out if you can control the node descriptor, as a device could be a router, but still battery operated, can it?

**Reply by martin3000** [reply]
It already was an end device with leaf stack.

**Reply by martin3000** [reply]
if in ZHA I go into my device, click on cluster maintenance, choose the basic cluster, select  "power source" attribute and click the "get" button, nothing happens. my device does not report the attribute values. do you have any idea why?

**Reply by MattWestb** [reply]
You can using the dev console in SS5 for see the debug message if you have enabling debug (also need comport for the console defined in the hardware config).

If you have it as one SED you need waking it up for it shall receiving the command from the coordinator.

**Reply by martin3000** [reply]
where can I configure whether the device is sleepy or not?

**Reply by MattWestb** [reply]
Zigbee, Stack and PRO Core. The leaf is for SED and normal stack is for routers.
I dont knowing if need more setting being changed then i have only making NCP and Zigbee Control Bridges.

**Reply by Adminiuga** [reply]
În New SDK it is under device config component.

Did you implement the power source attribute in your application? ZHA per se it is not using that attribute to tell if the device is battery operated or not.

**Reply by martin3000** [reply]
power source is set to battery, but home assistant ZHA always tell its mains powered.
![image](https://user-images.githubusercontent.com/21184101/148186788-dd5e92df-3b44-4bad-beed-c2de8e29e097.png)

**Reply by martin3000** [reply]
![Screenshot from 2022-01-05 09-42-15](https://user-images.githubusercontent.com/21184101/148187307-835f321a-174c-4034-a720-50ebb635bcb7.png)

ah ok thanks

**Reply by Adminiuga** [reply]
Zha is using node descriptor to tell if the device is mains powered or not.
It is not using the power source attribute.
